File: R/tools/file_io.R
import::here(plyr, 'rbind.fill')
import::here(readxl, 'read_excel')
import::here(file.path(wd, 'R', 'tools', 'list_tools.R'),
    'filter_list_for_match', .character_only=TRUE)


## Functions
## append_many_csv
## list_files
## join_many_csv
## read_excel_or_csv


#' Aggregate csv files by appending them rowwise
#' 
#' @description
#' Read all the csv files from a directory and append them into a single dataframe
#' 
#' @export
append_many_csv <- function(dir_path, sep=',', row_names=NULL, return_list=FALSE) {
    filenames <- list.files(dir_path, full.names=TRUE)

    dfs <- new.env()
    for (file in filenames) {
        df <- read.csv(
            file, sep=sep, row.names=row_names,
            header=TRUE, check.names=FALSE
        )

        # remove nulls
        df <- Filter(function(x) !all(is.na(x)), df)  # columns
        df <- df[rowSums(is.na(df)) != ncol(df), ]  # rows

        # repair colnames
        missing_colnames <- which(nchar(colnames(df))==0)
        num_missing <- length(missing_colnames)
        if (num_missing > 0) {
            colnames(df)[which(nchar(colnames(df))==0)] <- paste0("X", 1:num_missing)
        }

        # add filename
        cols <- colnames(df)
        df[['filename']] <- basename(file)
        df <- df[, c('filename', cols)]

        dfs[[file]] <- df
    }
    
    dfs <- as.list(dfs)

    if (return_list) {
        return(dfs)
    } else {
        combined <- do.call(rbind.fill, dfs)
        return(combined)  
    }
}


#' List all files with a specific extension
#' 
#' @description
#' This is a thin wrapper around [list.files()].
#' 
#' @references
#' \href{https://stackoverflow.com/questions/7187442/filter-a-vector-of-strings-based-on-string-matching}{StackOverflow post}
#' 
list_files <- function(dir_path, ext=NULL, recursive = TRUE) {
    all_files = list.files(dir_path, recursive = recursive, full.name=TRUE)

    if (!is.null(ext)) {
        # See: https://stackoverflow.com/questions/7187442/filter-a-vector-of-strings-based-on-string-matching
        return (all_files[tools::file_ext(all_files)==ext])
    } else {
        return (all_files)
    }
}


#' Aggregate csv files by joining them column-wise
#' 
#' @description
#' Read all the csv or tsv files from a directory and left join them into a single dataframe
#' 
#' @references
#' \href{https://stackoverflow.com/questions/5319839/read-multiple-csv-files-into-separate-all_reads-frames}{StackOverflow post}
#' 
join_many_csv <- function(
    paths, index_cols, value_cols=NULL,
    all.x=FALSE, all.y=FALSE, recursive=TRUE
) {

    # distinguish if paths is a directory or a list of files
    if (length(paths)==1) {
        if (dir.exists(paths)) {
             paths <- list_files(paths, recursive=recursive)
        }
    } else if (length(paths[file.exists(paths)]) == 0) {
        stop(paste("no files found!"))
    }

    # split into csv and tsv files
    csv_paths = filter_list_for_match(paths, 'csv')
    csv_list <- lapply(csv_paths, read.csv, sep=',')
    tsv_paths = filter_list_for_match(paths, 'tsv')
    tsv_list <- lapply(tsv_paths, read.csv, sep='\t') 
    df_list = c(csv_list, tsv_list)

    filenames = c(tools::file_path_sans_ext(basename(paths)))
    # Warning: column names ‘count.x’, ‘count.y’ are duplicated in the result
    # See: https://stackoverflow.com/questions/38603668/suppress-any-emission-of-a-particular-warning-message
    withCallingHandlers({
        all_data <- Reduce(
            function(...) merge(..., by=index_cols, all.x=all.x, all.y=all.y),
            lapply(df_list, "[", c(index_cols, value_cols))
        )
    }, warning = function(w) {
        # print(conditionMessage(w))
        if (startsWith(conditionMessage(w), "column names")) {
            invokeRestart( "muffleWarning" )
        }
    })
    
    # rename columns
    colnames(all_data) = c(
        index_cols,  # index_cols
        as.list(outer(value_cols, filenames, paste, sep='-'))  # suffix value_cols with filename
    )
    return(all_data)
}


#' switch case to read excel or csv based on the extension
#' 
#' @export
read_excel_or_csv <- function(filepath) {
    ext=tools::file_ext(filepath)
    if (ext == 'xlsx') {
        df <- read_excel(filepath, .name_repair = "unique_quiet")
    } else if (ext == 'csv') {
        df <- read.csv(filepath, header=TRUE, check.names=FALSE)
    } else {
        stop('Please enter a xlsx or csv file.')
    }
    return(df)
}

File: python/tools/file_io.py
import os
import pandas as pd
from typing import List, Union
from .list_tools import filter_list_for_match


def append_many_csv(dir_path: str, sep: str = ',', recursive: bool = False,
                    include_filepath: bool = True, return_list: bool = False,
                    na_strings: List[str] = None) -> Union[pd.DataFrame, dict]:
    """Read all CSV files from a directory and append them into a single DataFrame."""
    if recursive:
        filenames = []
        for root, dirs, files in os.walk(dir_path):
            for f in files:
                if f.endswith('.csv'):
                    filenames.append(os.path.join(root, f))
    else:
        filenames = [os.path.join(dir_path, f) for f in os.listdir(dir_path)
                     if f.endswith('.csv')]

    if not filenames:
        return {} if return_list else None

    dfs = {}
    for filepath in filenames:
        try:
            df = pd.read_csv(filepath, sep=sep, na_values=na_strings)

            # Remove all-NA columns and rows
            df = df.dropna(axis=1, how='all')
            df = df.dropna(axis=0, how='all')

            # Repair empty column names
            empty_cols = [i for i, col in enumerate(df.columns) if not col or col.startswith('Unnamed')]
            for i, idx in enumerate(empty_cols):
                cols = list(df.columns)
                cols[idx] = f'X{i+1}'
                df.columns = cols

            # Add filename info
            if include_filepath:
                rel_path = os.path.relpath(filepath, dir_path)
                df.insert(0, 'filepath', rel_path)
                df.insert(1, 'filename', os.path.basename(filepath))

            dfs[filepath] = df
        except Exception as e:
            print(f"Error reading {filepath}: {e}")

    if return_list:
        return dfs
    else:
        if dfs:
            combined = pd.concat(dfs.values(), ignore_index=True)
            return combined
        return None


def list_files(dir_path: str, ext: str = None, recursive: bool = True) -> List[str]:
    """List all files with a specific extension."""
    all_files = []

    if recursive:
        for root, dirs, files in os.walk(dir_path):
            for f in files:
                all_files.append(os.path.join(root, f))
    else:
        all_files = [os.path.join(dir_path, f) for f in os.listdir(dir_path)
                     if os.path.isfile(os.path.join(dir_path, f))]

    if ext:
        all_files = [f for f in all_files if f.endswith(f'.{ext}')]

    return all_files


def join_many_csv(paths: Union[str, List[str]], index_cols: List[str],
                  value_cols: List[str] = None, all_x: bool = False,
                  all_y: bool = False, recursive: bool = True) -> pd.DataFrame:
    """Read CSV/TSV files and left join them column-wise."""
    # Get file list
    if isinstance(paths, str):
        if os.path.isdir(paths):
            paths = list_files(paths, recursive=recursive)

    if not paths or not any(os.path.exists(p) for p in paths):
        raise ValueError("No files found!")

    # Split by file type
    csv_paths = filter_list_for_match(paths, 'csv')
    tsv_paths = filter_list_for_match(paths, 'tsv')

    df_list = []
    filenames = []

    for fp in csv_paths:
        df_list.append(pd.read_csv(fp, sep=','))
        filenames.append(os.path.splitext(os.path.basename(fp))[0])

    for fp in tsv_paths:
        df_list.append(pd.read_csv(fp, sep='\t'))
        filenames.append(os.path.splitext(os.path.basename(fp))[0])

    if not df_list:
        raise ValueError("No valid CSV/TSV files found!")

    # Merge all dataframes
    cols_to_use = index_cols + (value_cols or [])
    result = df_list[0][cols_to_use].copy()

    for i, df in enumerate(df_list[1:], 1):
        how = 'outer' if (all_x and all_y) else ('left' if all_x else ('right' if all_y else 'inner'))
        result = result.merge(df[cols_to_use], on=index_cols, how=how,
                              suffixes=('', f'_{filenames[i]}'))

    # Rename columns with filename suffixes
    if value_cols:
        new_cols = index_cols.copy()
        for val_col in value_cols:
            for fname in filenames:
                new_cols.append(f"{val_col}-{fname}")
        if len(result.columns) == len(new_cols):
            result.columns = new_cols

    return result


def read_excel_or_csv(filepath: str) -> pd.DataFrame:
    """Read Excel or CSV file based on extension."""
    ext = os.path.splitext(filepath)[1].lower()

    if ext == '.xlsx' or ext == '.xls':
        return pd.read_excel(filepath)
    elif ext == '.csv':
        return pd.read_csv(filepath)
    else:
        raise ValueError('Please enter a xlsx or csv file.')
